actionpack/lib/action_controller/metal/flash.rb

```diff
@@ -37,10 +37,15 @@ def add_flash_types(*types)
             request.flash[type]
           end
           helper_method(type) if respond_to?(:helper_method)
-
           self._flash_types += [type]
         end
       end
+
+      # Extend AbstractController::Base::action_methods to remove any added flash types
+      # from the set
+      def action_methods #:nodoc:
+        super - _flash_types.map(&:to_s).to_set
+      end
```

Comment:
It's important that `action_methods` is memoised, since it's called during every request.

We can copy the approach from the `UrlFor` module and use the same instance variable name, so that `clear_action_methods!` still works:

https://github.com/rails/rails/blob/a2cb0b23b7233371d8a2e0f843885dae330da2b7/actionpack/lib/abstract_controller/url_for.rb#L24-L30

```suggestion
      def action_methods #:nodoc:
        @action_methods ||= super - _flash_types.map(&:to_s).to_set
      end
```